lib/ssz_ex.ex

```diff
@@ -184,6 +192,145 @@ defmodule LambdaEthereumConsensus.SszEx do
     end
   end
 
+  defp encode_container(container, schemas) do
+    {fixed_size_values, fixed_length, variable_values} = analyze_schemas(container, schemas)
+
+    with {:ok, variable_parts} <- encode_schemas(variable_values),
+         offsets = calculate_offsets(variable_parts, fixed_length),
+         variable_length =
+           Enum.reduce(variable_parts, 0, fn part, acc -> byte_size(part) + acc end),
+         :ok <- check_length(fixed_length, variable_length),
+         {:ok, fixed_parts} <-
+           replace_offsets(fixed_size_values, offsets)
+           |> encode_schemas do
+      (fixed_parts ++ variable_parts)
+      |> Enum.join()
+      |> then(&{:ok, &1})
+    end
+  end
+
+  defp analyze_schemas(container, schemas) do
+    schemas
+    |> Enum.reduce({[], 0, []}, fn {key, schema},
+                                   {acc_fixed_size_values, acc_fixed_length, acc_variable_values} ->
+      value = Map.fetch!(container, key)
+
+      if variable_size?(schema) do
+        {[:offset | acc_fixed_size_values], @bytes_per_length_offset + acc_fixed_length,
+         [{value, schema} | acc_variable_values]}
+      else
+        {[{value, schema} | acc_fixed_size_values], acc_fixed_length + get_fixed_size(schema),
+         acc_variable_values}
+      end
+    end)
+  end
+
+  defp encode_schemas(tuple_values) do
+    Enum.map(tuple_values, fn {value, schema} -> encode(value, schema) end)
+    |> flatten_results()
+  end
+
+  defp calculate_offsets(variable_parts, fixed_length) do
+    {offsets, _} =
+      Enum.reduce(variable_parts, {[], 0}, fn element, {res, acc} ->
+        sum = fixed_length + acc
+        {[{sum, {:int, 32}} | res], byte_size(element) + acc}
+      end)
+
+    offsets
+  end
+
+  defp replace_offsets(fixed_size_values, offsets) do
+    {fixed_size_values, _} =
+      Enum.reduce(fixed_size_values, {[], offsets}, fn element,
+                                                       {acc_fixed_list, acc_offsets_list} ->
+        if element == :offset do
+          [offset | rest_offsets] = acc_offsets_list
+          {[offset | acc_fixed_list], rest_offsets}
+        else
+          {[element | acc_fixed_list], acc_offsets_list}
+        end
+      end)
```

Comment:
Great reduce! A couple of improvements:

1) Move the lambda to a separate `replace_offset` function, with two clauses:

```elixir
def replace_offset(:offset, {acc_fixed_lists, [offset | rest_offsets]})

def replace_offset(element, {acc_fixed_list, acc_offset_list})
```

Also, aren't we missing a reverse here? If that's the case, why isn't a test failing?

Reply:
it is not necessary because `analyze_schemas` also does not reverse `fixed_size_values` :sweat_smile: